### PR TITLE
Rework HaloInfo interface

### PR DESCRIPTION
**Pull request by csarofeen**
Trying to rework the HaloInfo interface a bit. When working on `contiguity.h` I realized it was challenging to separate `HaloInfo` from `GpuLower`. I don't know if there's value in being able to support: 
https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/contiguity.cpp#L52-L55
when we aren't in the lowering process. However, the only thing `HaloInfo` really takes to construct is the compute at map and the fusion. So I separated out the reliance on `GpuLower` in `HaloInfo`. I also realized there aren't any accesses to HaloInfo that are non-constant, so I removed the non-const interfaces from the public API.

The other thing I wanted to do was to try using `shared_ptr` to pass the information around just to be a bit safer on lifetime and ownership. I'm generally concerned with the way we use `GpuLower` we could be holding on to a reference that's not really safe to hold onto since our classes don't have ownership, `GpuLower` does. I don't think we have much chance to suffer from this today, but I think the pattern is safer.

The other thing I like about using a `shared_ptr` is it's easier to use a pattern I'm often looking for which is "Take this reference, or construct a new object." Specifically we sometimes share analysis tools in lowering when we're doing something like scheduling. At scheduling we either might be creating a new `ComputeAtMap`, or we might be using one that exists. So if we have a class that requires a `ComputeAtMap` we could either pass in a shared pointer, or create a new one.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp [hunk 1]
```diff
@@ -161,7 +154,9 @@ void HaloInfo::setRootAxisInfo(
   return;
 }
 
-void HaloInfo::build(Fusion* fusion) {
+HaloInfo::HaloInfo(Fusion* fusion, std::shared_ptr<const ComputeAtMap> ca_map)
+    // Make a copy of the permissive map for extent comparators
+    : permissive_map_(ca_map->idGraph().permissiveNodes()) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_shift.cpp:159]: Are you trying to decouple the states held by `GpuLower` from each other? I like it as that seems like a better and more robust design.

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp [hunk 2]
```diff
@@ -801,7 +796,8 @@ std::unordered_map<IterDomain*, Val*> HaloInfo::buildConcreteHaloExtentMap(
         merged_shifted_ids.insert(ir_utils::caMapExactConcreteId(merge->out()));
         // Note that halo_width_map_ is not updated
       } else {
-        setHaloWidth(ir_utils::caMapExactConcreteId(merge->out()), 0);
+        local_halo_info.setHaloWidth(
+            ir_utils::caMapExactConcreteId(merge->out()), 0);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_shift.cpp:800]: Regarding the decoupling, here we still use the CA map held by GpuLower. And, also just realized it's a little counter-intuitive that `ir_utils` implicitly uses `GpuLower`. It doesn't seem trivial to remove this interface as it's used everywhere.

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the cleanup.

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:159]: Yes, I'm trying to pull the dependencies apart a bit as I was thinking of building a HaloInfo structure in contiguity which is also used outside GpuLower.

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:800]: Is making this `local_halo_info` right?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:800]: Will look again at the inter-class dependencies.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:800]: Oh, this class is really hard set to the lowering process as it's taking indexing in. That's why I didn't bother looking more closely.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:800]: `ir_utils::caMapExactConcreteId` seems like a weak alias, will get rid of that either before this is merged or in another PR.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:800]: Creating `local_halo_info` is right. It was part of Shiming's indexing refactoring. Previously, reference domains are added to the singleton instance of `HaloInfo`, but as we don't create reference IterDomains anymore, the properties generated for reference IterDomains are instead generated for concrete IterDomains. A complexity about halo is that even exact mapped IterDomains may have different halo sizes, so it's not possible to associate a unique halo attribute with a group of mapped IterDomains. As the halo size information depends on the loop nest where an expression appears, what Shiming did was to clone the pre-built HaloInfo and modify the halo information for the used concrete IterDomains at that loop nest. 

That's my understanding from my reading of his refactoring PR. Tagging @shmsong as he may want to chime in.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp:800]: Yeah, I was assuming it was just a typo.

**Pull request merged**